File: Java/105. Construct Binary Tree from Preorder and Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    Map<Integer, Integer> mapping;
    // use this variable to record the root's index
    int preIndex;
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        preIndex = 0;
        mapping = new HashMap<>();
        // add the mapping between value and its index for quick lookup
        for(int i=0; i<inorder.length; i++){
            mapping.put(inorder[i], i);
        }
        
        return arrayTree(preorder, 0, preorder.length-1);
        
    }
    
    private TreeNode arrayTree(int[] preorder, int left, int right){
        // visited all nodes
        if(left > right){
            return null;
        }
        
        // get the root's value and its index in inorder
        int value = preorder[preIndex++];
        int index = mapping.get(value);
            
        TreeNode root = new TreeNode(value);
        root.left = arrayTree(preorder, left, index-1);
        root.right = arrayTree(preorder, index+1, right);
        
        return root;
    }
}

File: Python3/105. Construct Binary Tree from Preorder and Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder:
            return None
        # the first element in preorder is the root node
        value = preorder[0]
        
        index = inorder.index(value)
        #print(value,index)
        
        root = TreeNode(value)
        root.left = self.buildTree(preorder[1:1+index], inorder[:index])
        root.right = self.buildTree(preorder[1+index:], inorder[index+1:])
        
        return root

# Solution 2
# O(N) time and O(N) space
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder:
            return None
        
        mapping = {}
        for i in range(len(inorder)):
            mapping[inorder[i]] = i
        
        
        def arrayTree(left, right):
            if left > right:
                return None
            
            # the first element in preorder is the root node
            value = preorder.pop(0)
            
            # get the index of the root through the dict, O(1)
            index = mapping[value]
            #print(value,index)

            root = TreeNode(value)
            root.left = arrayTree(left, index-1)
            root.right = arrayTree(index+1, right)
        
            return root
        
        return arrayTree(0, len(preorder)-1)